clippy_lints/src/hashset_insert_after_contains.rs

```diff
@@ -0,0 +1,137 @@
+use std::ops::ControlFlow;
+
+use clippy_utils::diagnostics::span_lint_and_then;
+use clippy_utils::ty::is_type_diagnostic_item;
+use clippy_utils::visitors::for_each_expr;
+use clippy_utils::{higher, peel_hir_expr_while, SpanlessEq};
+use rustc_hir::{Expr, ExprKind, UnOp};
+use rustc_lint::{LateContext, LateLintPass};
+use rustc_session::declare_lint_pass;
+use rustc_span::{sym, Span};
+
+declare_clippy_lint! {
+    /// ### What it does
+    /// Checks for usage of `contains` to see if a value is not
+    /// present on `HashSet` followed by a `insert`.
+    ///
+    /// ### Why is this bad?
+    /// Using just `insert` and checking the returned `bool` is more efficient.
+    ///
+    /// ### Example
+    /// ```rust
+    /// use std::collections::HashSet;
+    /// let mut set = HashSet::new();
+    /// let value = 5;
+    /// if !set.contains(&value) {
+    ///     set.insert(value);
+    ///     println!("inserted {value:?}");
+    /// }
+    /// ```
+    /// Use instead:
+    /// ```rust
+    /// use std::collections::HashSet;
+    /// let mut set = HashSet::new();
+    /// let value = 5;
+    /// if set.insert(&value) {
+    ///     println!("inserted {value:?}");
+    /// }
+    /// ```
+    #[clippy::version = "1.80.0"]
+    pub HASHSET_INSERT_AFTER_CONTAINS,
+    nursery,
+    "unnecessary call to `HashSet::contains` followed by `HashSet::insert`"
+}
+
+declare_lint_pass!(HashsetInsertAfterContains => [HASHSET_INSERT_AFTER_CONTAINS]);
+
+impl<'tcx> LateLintPass<'tcx> for HashsetInsertAfterContains {
+    fn check_expr(&mut self, cx: &LateContext<'tcx>, expr: &'tcx Expr<'_>) {
+        if !expr.span.from_expansion()
+            && let Some(higher::If {
+                cond: cond_expr,
+                then: then_expr,
+                ..
+            }) = higher::If::hir(expr)
+            && let Some(contains_expr) = try_parse_contains(cx, cond_expr)
+            && find_insert_calls(cx, &contains_expr, then_expr)
+        {
+            span_lint_and_then(
+                cx,
+                HASHSET_INSERT_AFTER_CONTAINS,
+                expr.span,
+                "usage of `HashSet::insert` after `HashSet::contains`",
+                |diag| {
+                    diag.note("`HashSet::insert` returns whether it was inserted")
+                        .span_help(contains_expr.span, "remove the `HashSet::contains` call");
+                },
+            );
+        }
+    }
+}
+
+struct ContainsExpr<'tcx> {
+    receiver: &'tcx Expr<'tcx>,
+    value: &'tcx Expr<'tcx>,
+    span: Span,
+}
+fn try_parse_contains<'tcx>(cx: &LateContext<'_>, expr: &'tcx Expr<'_>) -> Option<ContainsExpr<'tcx>> {
```

Comment:
`try_parse_insert` and `try_parse_contains` are equal on all but two things: The `UnOp` (Not vs. Deref) and the call sym (`insert` vs. `contains`). Please factor out a `try_parse_op_call` method to use in both cases (oh, and the `contains` result also containing the span, but we can live with ignoring that for the `insert` case later on).

Reply:
@llogiq thanks for your comment. I was trying to take a look at it and I just have one doubt that I wanted to validate with you.

There is another difference betwheen `try_parse_insert` and `try_parse_contains`: on `try_parse_insert` I am peeling the `expr` while it is a `Not`, at the beginning, and at the `try_parse_contains` I am peeling the `value` of the `expr` - if it is a method call` - while it is a `Deref`. So the `UnOp` is applied on different things (`expr` vs value of `expr`). To make a generic function I think that I would need to always do both peelings. I don't think it would be a problem, in terms of correctness, but probably it is not necessary, so I just wanted to double check with you.

Does it make sense?

Reply:
Great, did it! https://github.com/rust-lang/rust-clippy/pull/12873/commits/2b0cad6451cfe192e251379b741e1214cb36623a